packages/teuchos/parameterlist/src/Teuchos_ParameterListModifier.cpp

```diff
@@ -0,0 +1,213 @@
+    const Array<std::string> &exclude_sublists) const
+{
+  int replacements = 0;
+  auto ignore_names = exclude_sublists;
+  std::sort(ignore_names.begin(), ignore_names.end());
+  ParameterList::ConstIterator itr;
+  if (valid_pl.isSublist(sublist_template_name)){
+    ParameterList &valid_pl_sublist = valid_pl.get<ParameterList>(sublist_template_name);
+    for (itr = pl.begin(); itr != pl.end(); ++itr) {
+      const std::string &subname = pl.name(itr);
+      if (pl.isSublist(subname)){
+        if (!std::binary_search(ignore_names.begin(), ignore_names.end(), subname)){
+          valid_pl.set(subname, valid_pl_sublist);
+          replacements += 1;
+        }
+      }
+    }
+    valid_pl.remove(sublist_template_name);
+  }
+  return replacements;
+}
+
+
+int ParameterListModifier::expandSublistsUsingBaseName(
+    const std::string &base_name, ParameterList &pl, ParameterList &valid_pl,
+    const bool &allow_base_name) const
+{
+  int replacements = 0;
+  bool delete_base_name = true;
+  if (valid_pl.isSublist(base_name)){
+    if (pl.isSublist(base_name)){
+      if (!allow_base_name){
+        throw std::logic_error("Sublist can't have the same name as the parameter template name without `allow_base_name=true`.");
+      } else{
+        delete_base_name = false;
+      }
+    }
+    Array<std::string> matches = findMatchingBaseNames(pl, base_name, false, true);
+    replacements = matches.length();
+    for (const std::string &match_name : matches){
+      valid_pl.set(match_name, valid_pl.get<ParameterList>(base_name));
+    }
+    if (delete_base_name){
+      valid_pl.remove(base_name);
+    }
+  }
+  return replacements;
+}
+
+
+int ParameterListModifier::setDefaultsInSublists(const std::string &param_name, ParameterList &pl,
+    const Array<std::string> &sublist_names) const{
+  int num_defaults = 0;
+  if (pl.isParameter(param_name)){
+    for (const std::string &sublist_name : sublist_names){
+      if (pl.isSublist(sublist_name)){
+        ParameterList &sublist = pl.sublist(sublist_name);
+        if (!sublist.isParameter(param_name)){
+          ParameterEntry &pl_entry = pl.getEntry(param_name);
+          sublist.setEntry(param_name, pl_entry);
+        }
+      }
+    }
+    pl.remove(param_name);
+  }
+  return num_defaults;
+}
+
+
+template<typename T>
+bool ParameterListModifier::replaceScalarParameterWithArray(const std::string &param_name, const std::string &new_name,
+    ParameterList &pl, const bool &throw_if_new_name_exists) const{
+  bool param_exists = false;
+  if (pl.isParameter(param_name)){
+    param_exists = true;
+    if (!pl.isType<T>(param_name)){
+      throw std::logic_error("The parameter " + new_name + " is not of type " + typeid(T).name());
+    }
+    if ((pl.isParameter(new_name) || pl.isSublist(new_name)) && throw_if_new_name_exists){
+      throw std::logic_error("The parameter " + new_name + " already exists in this parameter list.");
```

Comment:
Please use `TEUCHOS_TEST_FOR_EXCEPTION()` instead of raw `throw`.

Reply:
Fixed in next push